flow/layers/display_list_layer.cc

```diff
@@ -0,0 +1,130 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/flow/layers/display_list_layer.h"
+
+#include "flutter/flow/display_list_canvas.h"
+
+namespace flutter {
+
+DisplayListLayer::DisplayListLayer(const SkPoint& offset,
+                                   sk_sp<DisplayList> display_list,
+                                   bool is_complex,
+                                   bool will_change)
+    : offset_(offset),
+      display_list_(display_list),
+      is_complex_(is_complex),
+      will_change_(will_change) {}
+
+#ifdef FLUTTER_ENABLE_DIFF_CONTEXT
+
+bool DisplayListLayer::IsReplacing(DiffContext* context,
+                                   const Layer* layer) const {
+  // Only return true for identical display lists; This way
+  // ContainerLayer::DiffChildren can detect when a display list layer
+  // got inserted between other display list layers
+  auto old_layer = layer->as_display_list_layer();
+  return old_layer != nullptr && offset_ == old_layer->offset_ &&
+         Compare(context->statistics(), this, old_layer);
+}
+
+void DisplayListLayer::Diff(DiffContext* context, const Layer* old_layer) {
+  DiffContext::AutoSubtreeRestore subtree(context);
+  if (!context->IsSubtreeDirty()) {
+#ifndef NDEBUG
+    FML_DCHECK(old_layer);
+    auto prev = old_layer->as_display_list_layer();
+    DiffContext::Statistics dummy_statistics;
+    // IsReplacing has already determined that the display list is same
+    FML_DCHECK(prev->offset_ == offset_ &&
+               Compare(dummy_statistics, this, prev));
+#endif
+  }
+  context->PushTransform(SkMatrix::Translate(offset_.x(), offset_.y()));
+  context->AddLayerBounds(display_list_->bounds());
+  context->SetLayerPaintRegion(this, context->CurrentSubtreeRegion());
+}
+
+bool DisplayListLayer::Compare(DiffContext::Statistics& statistics,
+                               const DisplayListLayer* l1,
+                               const DisplayListLayer* l2) {
+  const auto& dl1 = l1->display_list_;
+  const auto& dl2 = l2->display_list_;
+  if (dl1.get() == dl2.get()) {
+    statistics.AddSameInstancePicture();
+    return true;
+  }
+  auto op_cnt_1 = dl1->opCount();
+  auto op_cnt_2 = dl2->opCount();
+  if (op_cnt_1 != op_cnt_2 || dl1->bounds() != dl2->bounds()) {
+    statistics.AddNewPicture();
+    return false;
+  }
+
+  if (op_cnt_1 > 10) {
+    statistics.AddPictureTooComplexToCompare();
+    return false;
+  }
```

Comment:
Why 10?

Could we somehow surface in the display list interface whether it has expensive-to-compare ops in it? For example, if there's a single op that takes 100ns to compare, but 50 ops that each take 1ns...

Do we have benchmarks to say that this is a good number to go with? If not, do we have a plan (and an issue filed) to make this more reliable/performant?

Reply:
I copied that from the PictureLayer implementation for now. It can be tuned over time. We might want to have a higher threshold given that the DL.equals() method exists and is much more efficient than the SkPicture technique of serializing the stream and comparing a hash.

And for now this is based on the count of the records in the DL. We could also compute and base this on a complexity metric. Since the vast majority of the ops are compared using a bulk compare, we could base it on the number of bytes. I think @knopp probably has some experience with how to tune this metric. For this first pass, I should probably bump this up or eliminate the test - Matt?

Reply:
I switched to a byte size comparison with a guess at a threshold of 10,000 bytes. I am waiting to hear back from @knopp before I push the change.